<a href="https://colab.research.google.com/github/gayathrig21/Miscellaneous/blob/main/INSAID_Practise_EDA_Day16.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Sample "tips" dataset in seaborn

In [1]:
import seaborn as sns
import pandas as pd
tips= sns.load_dataset('tips')
tips.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

In [2]:
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


Info gives no of rows 244 , no of columns 7 and no missing data

In [3]:
tips.describe()

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

Gives statistical information.

Total bill : Mean > Median (positive skewness)

Size : Integer column cannot have float numbers as it denotes no of persons

Outliers in all the columns



In [4]:
tips.describe(include="all")

total_bill         tip   sex smoker  day    time        size
count   244.000000  244.000000   244    244  244     244  244.000000
unique         NaN         NaN     2      2    4       2         NaN
top            NaN         NaN  Male     No  Sat  Dinner         NaN
freq           NaN         NaN   157    151   87     176         NaN
mean     19.785943    2.998279   NaN    NaN  NaN     NaN    2.569672
std       8.902412    1.383638   NaN    NaN  NaN     NaN    0.951100
min       3.070000    1.000000   NaN    NaN  NaN     NaN    1.000000
25%      13.347500    2.000000   NaN    NaN  NaN     NaN    2.000000
50%      17.795000    2.900000   NaN    NaN  NaN     NaN    2.000000
75%      24.127500    3.562500   NaN    NaN  NaN     NaN    3.000000
max      50.810000   10.000000   NaN    NaN  NaN     NaN    6.000000

include = "all" gives categorical columns

It gives information about unique values for categorical column

top values  - (Mode)

**Inference** :

More non smokers  

More male 

More footfall on saturday 

More transactions for dinner 






In [5]:
tips['sex'].unique()

['Female', 'Male']
Categories (2, object): ['Male', 'Female']

Do not assume , check the data and be sure

In [6]:
tips['time'].unique()

['Dinner', 'Lunch']
Categories (2, object): ['Lunch', 'Dinner']

In [7]:
tips['time'].value_counts()

Dinner    176
Lunch      68
Name: time, dtype: int64

In [8]:
print(tips[tips['time']=='Lunch']['tip'].sum())
print(tips[tips['time']=='Dinner']['tip'].sum())

185.51000000000002
546.0699999999999


we cannot infer from this we need to know the mean

In [9]:
print('Lunch Mean', tips[tips['time']=='Lunch']['tip'].mean())
print('Dinner Mean',tips[tips['time']=='Dinner']['tip'].mean())
print('Lunch Median', tips[tips['time']=='Lunch']['tip'].median())
print('Dinner Median',tips[tips['time']=='Dinner']['tip'].median())

Lunch Mean 2.728088235294118
Dinner Mean 3.102670454545454
Lunch Median 2.25
Dinner Median 3.0


Looking at Mean and Median we can infer that Dinner tip is more consistent than lunch tip

In [10]:
tips.groupby(['time']).sum()

total_bill     tip  size
time                            
Lunch      1167.47  185.51   164
Dinner     3660.30  546.07   463

In [11]:
tips.groupby(['time']).mean()

total_bill       tip      size
time                                  
Lunch    17.168676  2.728088  2.411765
Dinner   20.797159  3.102670  2.630682

In [12]:
tips.groupby(['time']).count()

total_bill  tip  sex  smoker  day  size
time                                           
Lunch           68   68   68      68   68    68
Dinner         176  176  176     176  176   176

In [13]:
tips_numerical=tips.drop(columns=['sex', 'smoker', 'day'], axis=1)

In [14]:
tips_numerical.groupby(['time']).agg(['sum','mean','median','count'])

total_bill                              tip                         \
              sum       mean  median count     sum      mean median count   
time                                                                        
Lunch     1167.47  17.168676  15.965    68  185.51  2.728088   2.25    68   
Dinner    3660.30  20.797159  18.390   176  546.07  3.102670   3.00   176   

       size                         
        sum      mean median count  
time                                
Lunch   164  2.411765    2.0    68  
Dinner  463  2.630682    2.0   176

Aggregate all the values at once

In [15]:
tips.groupby(['time','sex']).agg(['sum','mean','median','count'])

<ipython-input-15-9d17d76cfebb>:1: FutureWarning: ['smoker', 'day'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.
  tips.groupby(['time','sex']).agg(['sum','mean','median','count'])


total_bill                             tip                   \
                     sum       mean median count     sum      mean median   
time   sex                                                                  
Lunch  Male       595.60  18.048485  16.58    33   95.11  2.882121   2.31   
       Female     571.87  16.339143  13.42    35   90.40  2.582857   2.01   
Dinner Male      2661.22  21.461452  19.63   124  389.96  3.144839   3.00   
       Female     999.08  19.213077  17.19    52  156.11  3.002115   3.00   

                    size                         
              count  sum      mean median count  
time   sex                                       
Lunch  Male      33   78  2.363636    2.0    33  
       Female    35   86  2.457143    2.0    35  
Dinner Male     124  335  2.701613    2.0   124  
       Female    52  128  2.461538    2.0    52

In [16]:
tips.groupby(['day','time','sex'])['tip'].agg(['mean','median'])

mean  median
day  time   sex                     
Thur Lunch  Male    2.980333   2.530
            Female  2.561935   2.000
     Dinner Male         NaN     NaN
            Female  3.000000   3.000
Fri  Lunch  Male    1.900000   1.920
            Female  2.745000   2.750
     Dinner Male    3.032857   3.000
            Female  2.810000   3.000
Sat  Lunch  Male         NaN     NaN
            Female       NaN     NaN
     Dinner Male    3.083898   3.000
            Female  2.801786   2.625
Sun  Lunch  Male         NaN     NaN
            Female       NaN     NaN
     Dinner Male    3.220345   3.085
            Female  3.367222   3.500

**Saturday** and **Sunday** no lunch is served


Aggregate funs: Mean , median, max, min , sd, range 

**`Time series data`** - Single dimension data

In [17]:
from google.colab import files
uploaded = files.upload()

Saving kaggle.json to kaggle (1).json


In [18]:
!chmod 600 /content/kaggle.json

In [19]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [20]:
!pwd

/content


In [21]:
import os
os.environ['KAGGLE_CONFIG_DIR']='/content'

In [22]:
!kaggle datasets download -d fedesoriano/air-quality-data-set

air-quality-data-set.zip: Skipping, found more recently modified local copy (use --force to force download)


In [23]:
!pip install patool

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [24]:
import patoolib

In [25]:
patoolib.extract_archive('air-quality-data-set.zip')

patool: Extracting air-quality-data-set.zip ...
patool: running /usr/bin/7z x -o./Unpack_m17vbzc8 -- air-quality-data-set.zip
patool: ... air-quality-data-set.zip extracted to `air-quality-data-set' (local file exists).


'air-quality-data-set'

In [26]:
air_df=pd.read_csv('AirQuality.csv',sep=';')

In [27]:
air_df.head()

Date      Time CO(GT)  PT08.S1(CO)  NMHC(GT) C6H6(GT)  PT08.S2(NMHC)  \
0  10/03/2004  18.00.00    2,6       1360.0     150.0     11,9         1046.0   
1  10/03/2004  19.00.00      2       1292.0     112.0      9,4          955.0   
2  10/03/2004  20.00.00    2,2       1402.0      88.0      9,0          939.0   
3  10/03/2004  21.00.00    2,2       1376.0      80.0      9,2          948.0   
4  10/03/2004  22.00.00    1,6       1272.0      51.0      6,5          836.0   

   NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)     T    RH  \
0    166.0        1056.0    113.0        1692.0       1268.0  13,6  48,9   
1    103.0        1174.0     92.0        1559.0        972.0  13,3  47,7   
2    131.0        1140.0    114.0        1555.0       1074.0  11,9  54,0   
3    172.0        1092.0    122.0        1584.0       1203.0  11,0  60,0   
4    131.0        1205.0    116.0        1490.0       1110.0  11,2  59,6   

       AH  Unnamed: 15  Unnamed: 16  
0  0,7578          NaN          NaN  
1  0,7255          NaN          NaN  
2  0,7502          NaN          NaN  
3  0,7867          NaN          NaN  
4  0,7888          NaN          NaN

In [28]:
type(air_df['CO(GT)'])

pandas.core.series.Series

In [29]:
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         9357 non-null   object 
 3   PT08.S1(CO)    9357 non-null   float64
 4   NMHC(GT)       9357 non-null   float64
 5   C6H6(GT)       9357 non-null   object 
 6   PT08.S2(NMHC)  9357 non-null   float64
 7   NOx(GT)        9357 non-null   float64
 8   PT08.S3(NOx)   9357 non-null   float64
 9   NO2(GT)        9357 non-null   float64
 10  PT08.S4(NO2)   9357 non-null   float64
 11  PT08.S5(O3)    9357 non-null   float64
 12  T              9357 non-null   object 
 13  RH             9357 non-null   object 
 14  AH             9357 non-null   object 
 15  Unnamed: 15    0 non-null      float64
 16  Unnamed: 16    0 non-null      float64
dtypes: float64(10), object(7)
memory usage: 1.2+ MB


In [30]:
air_df.drop(columns=['Unnamed: 15','Unnamed: 16'], axis=1 , inplace=True)

In [31]:
air_df.describe()

PT08.S1(CO)     NMHC(GT)  PT08.S2(NMHC)      NOx(GT)  PT08.S3(NOx)  \
count  9357.000000  9357.000000    9357.000000  9357.000000   9357.000000   
mean   1048.990061  -159.090093     894.595276   168.616971    794.990168   
std     329.832710   139.789093     342.333252   257.433866    321.993552   
min    -200.000000  -200.000000    -200.000000  -200.000000   -200.000000   
25%     921.000000  -200.000000     711.000000    50.000000    637.000000   
50%    1053.000000  -200.000000     895.000000   141.000000    794.000000   
75%    1221.000000  -200.000000    1105.000000   284.000000    960.000000   
max    2040.000000  1189.000000    2214.000000  1479.000000   2683.000000   

           NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  
count  9357.000000   9357.000000  9357.000000  
mean     58.148873   1391.479641   975.072032  
std     126.940455    467.210125   456.938184  
min    -200.000000   -200.000000  -200.000000  
25%      53.000000   1185.000000   700.000000  
50%      96.000000   1446.000000   942.000000  
75%     133.000000   1662.000000  1255.000000  
max     340.000000   2775.000000  2523.000000

In [32]:
missing_nmhc=air_df[air_df['NMHC(GT)']==-200]['NMHC(GT)'].count()

In [33]:
percent_missing= (missing_nmhc/air_df['NMHC(GT)'].count())*100

In [34]:
percent_missing

90.23191193758683

Since 90% of the data in the column 'NMHC' is missing we can drop the column

In [35]:
air_df.drop(columns=['NMHC(GT)'], axis=1, inplace=True)

In [38]:
air_df['Date']=pd.to_datetime(air_df['Date'], format='%d/%m/%Y')
air_df['Time']=pd.to_datetime(air_df['Time'], format='%H.%M.%S')

In [39]:
air_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9471 entries, 0 to 9470
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date           9357 non-null   datetime64[ns]
 1   Time           9357 non-null   datetime64[ns]
 2   CO(GT)         9357 non-null   object        
 3   PT08.S1(CO)    9357 non-null   float64       
 4   C6H6(GT)       9357 non-null   object        
 5   PT08.S2(NMHC)  9357 non-null   float64       
 6   NOx(GT)        9357 non-null   float64       
 7   PT08.S3(NOx)   9357 non-null   float64       
 8   NO2(GT)        9357 non-null   float64       
 9   PT08.S4(NO2)   9357 non-null   float64       
 10  PT08.S5(O3)    9357 non-null   float64       
 11  T              9357 non-null   object        
 12  RH             9357 non-null   object        
 13  AH             9357 non-null   object        
dtypes: datetime64[ns](2), float64(7), object(5)
memory usage: 1.0+ MB


In [40]:
air_df['Date'].unique()

array(['2004-03-10T00:00:00.000000000', '2004-03-11T00:00:00.000000000',
       '2004-03-12T00:00:00.000000000', '2004-03-13T00:00:00.000000000',
       '2004-03-14T00:00:00.000000000', '2004-03-15T00:00:00.000000000',
       '2004-03-16T00:00:00.000000000', '2004-03-17T00:00:00.000000000',
       '2004-03-18T00:00:00.000000000', '2004-03-19T00:00:00.000000000',
       '2004-03-20T00:00:00.000000000', '2004-03-21T00:00:00.000000000',
       '2004-03-22T00:00:00.000000000', '2004-03-23T00:00:00.000000000',
       '2004-03-24T00:00:00.000000000', '2004-03-25T00:00:00.000000000',
       '2004-03-26T00:00:00.000000000', '2004-03-27T00:00:00.000000000',
       '2004-03-28T00:00:00.000000000', '2004-03-29T00:00:00.000000000',
       '2004-03-30T00:00:00.000000000', '2004-03-31T00:00:00.000000000',
       '2004-04-01T00:00:00.000000000', '2004-04-02T00:00:00.000000000',
       '2004-04-03T00:00:00.000000000', '2004-04-04T00:00:00.000000000',
       '2004-04-05T00:00:00.000000000', '2004-04-06

In [41]:
air_df['Date'].dt.date.unique()

array([datetime.date(2004, 3, 10), datetime.date(2004, 3, 11),
       datetime.date(2004, 3, 12), datetime.date(2004, 3, 13),
       datetime.date(2004, 3, 14), datetime.date(2004, 3, 15),
       datetime.date(2004, 3, 16), datetime.date(2004, 3, 17),
       datetime.date(2004, 3, 18), datetime.date(2004, 3, 19),
       datetime.date(2004, 3, 20), datetime.date(2004, 3, 21),
       datetime.date(2004, 3, 22), datetime.date(2004, 3, 23),
       datetime.date(2004, 3, 24), datetime.date(2004, 3, 25),
       datetime.date(2004, 3, 26), datetime.date(2004, 3, 27),
       datetime.date(2004, 3, 28), datetime.date(2004, 3, 29),
       datetime.date(2004, 3, 30), datetime.date(2004, 3, 31),
       datetime.date(2004, 4, 1), datetime.date(2004, 4, 2),
       datetime.date(2004, 4, 3), datetime.date(2004, 4, 4),
       datetime.date(2004, 4, 5), datetime.date(2004, 4, 6),
       datetime.date(2004, 4, 7), datetime.date(2004, 4, 8),
       datetime.date(2004, 4, 9), datetime.date(2004, 4, 10),
 

In [42]:
air_df['Date'].dt.day.unique()

array([10., 11., 12., 13., 14., 15., 16., 17., 18., 19., 20., 21., 22.,
       23., 24., 25., 26., 27., 28., 29., 30., 31.,  1.,  2.,  3.,  4.,
        5.,  6.,  7.,  8.,  9., nan])

In [43]:
air_df['Date'].dt.year.unique()

array([2004., 2005.,   nan])

In [44]:
air_df['Time'].dt.hour.unique()

array([18., 19., 20., 21., 22., 23.,  0.,  1.,  2.,  3.,  4.,  5.,  6.,
        7.,  8.,  9., 10., 11., 12., 13., 14., 15., 16., 17., nan])

In [45]:
from datetime import timedelta
air_df['DateNew']=air_df['Date']+timedelta(days=2)

In [46]:
air_df

Date                Time CO(GT)  PT08.S1(CO) C6H6(GT)  \
0    2004-03-10 1900-01-01 18:00:00    2,6       1360.0     11,9   
1    2004-03-10 1900-01-01 19:00:00      2       1292.0      9,4   
2    2004-03-10 1900-01-01 20:00:00    2,2       1402.0      9,0   
3    2004-03-10 1900-01-01 21:00:00    2,2       1376.0      9,2   
4    2004-03-10 1900-01-01 22:00:00    1,6       1272.0      6,5   
...         ...                 ...    ...          ...      ...   
9466        NaT                 NaT    NaN          NaN      NaN   
9467        NaT                 NaT    NaN          NaN      NaN   
9468        NaT                 NaT    NaN          NaN      NaN   
9469        NaT                 NaT    NaN          NaN      NaN   
9470        NaT                 NaT    NaN          NaN      NaN   

      PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  \
0            1046.0    166.0        1056.0    113.0        1692.0   
1             955.0    103.0        1174.0     92.0        1559.0   
2             939.0    131.0        1140.0    114.0        1555.0   
3             948.0    172.0        1092.0    122.0        1584.0   
4             836.0    131.0        1205.0    116.0        1490.0   
...             ...      ...           ...      ...           ...   
9466            NaN      NaN           NaN      NaN           NaN   
9467            NaN      NaN           NaN      NaN           NaN   
9468            NaN      NaN           NaN      NaN           NaN   
9469            NaN      NaN           NaN      NaN           NaN   
9470            NaN      NaN           NaN      NaN           NaN   

      PT08.S5(O3)     T    RH      AH    DateNew  
0          1268.0  13,6  48,9  0,7578 2004-03-12  
1           972.0  13,3  47,7  0,7255 2004-03-12  
2          1074.0  11,9  54,0  0,7502 2004-03-12  
3          1203.0  11,0  60,0  0,7867 2004-03-12  
4          1110.0  11,2  59,6  0,7888 2004-03-12  
...           ...   ...   ...     ...        ...  
9466          NaN   NaN   NaN     NaN        NaT  
9467          NaN   NaN   NaN     NaN        NaT  
9468          NaN   NaN   NaN     NaN        NaT  
9469          NaN   NaN   NaN     NaN        NaT  
9470          NaN   NaN   NaN     NaN        NaT  

[9471 rows x 15 columns]